Template Haskell quotes as patterns

Ericson2314 opened the pull request:
The Template Haskell AST is inherently unstable. Let's make that less of an issue. We can do that by allowing Quotes to be used instead of Template Haskell AST data constructors in patterns.

david-christiansen commented:
A few questions about the interpretation of quotation patterns:

Would you allow any of the following, and would they recognize precisely the lambda expressions?
1. `isLambda [| \_ -> $(_) |] = True`
2. `isLambda [| \$(_) -> $(_) |] = True`

Is `isLambdaLambda [| \$(_) -> $(_) -> $(_) |] = True` equivalent to `isLambdaLambda [| \$(_) $(_) -> $(_) |] = True`?

Given that the language in the module that invokes the metaprogram that uses a quotation pattern may be using different GHC extensions than the module in which it occurs, what happens if there's e.g. `BlockArguments` enabled in just one of them? Are there any sets of incompatible syntactic extensions that would make it impossible to write a metaprogram that matched all possible user syntax?

Ericson2314 commented:
> `isLambda [| \$(_) -> $(_) |] = True`

Returns `True` for any single parameter lambda. This is allowed.

> `isLambda [| \_ -> $(_) |] = True`

Returns `True` for any lambda with a single `_` parameter. Not the same!

> Is `isLambdaLambda [| \$(_) -> $(_) -> $(_) |] = True` equivalent to `isLambdaLambda [| \$(_) $(_) -> $(_) |] = True`?

I assume you meant ``isLambdaLambda [| \$(_) -> \$(_) -> $(_) |] = True` for the first one? (Note the second `\` I added.)

Those would *not* be equivalent because the syntax being matched is different even though syntax has the same semantics.

(It would be interesting to do macros that respect various equational reasoning on syntax, but that is a huge leap from what we have today with Template Haskell. I think it is better to leave that out of scope.)

------

Instead, I rather create new splicing forms that allow one to e.g. splice (in positive and negative position) a list of parameters. e.g. suppose we could could do
```haskell
isLambda [| \$(x) $...(_) -> $(_) |] = True
```
and that would match both
```haskell
\(Pat foo) -> ignored
\(Pat foo) ignored -> ignored
```
in both cases binding `[p| Pat foo |]` as `x`.

My brief experience with @int-index exploring converting `happy` to TH (https://github.com/haskell/happy/tree/wip/template-haskell) indicated to me that were *greatly* lacking expressive power in the sorts of quotes and slices that are permitted --- the `$...(thing)` example being just one example.

Still, this too I think is best left as future work. We can nail down the core feature now, and then with a good foundation open the floodgates to much more quoting/splicing flexibility, I think that will work better.

TeofilC commented on proposals/0529-th-quotes-as-patterns.rst:
```diff
@@ -0,0 +1,344 @@
+Scheme and Racket
+~~~~~~~~~~~~~~~~~
+
+- ``syntax-rules``, see https://docs.racket-lang.org/guide/syntax-rules.html .
```
This link looks to be dead

Ericson2314 replied:
```suggestion
- ``syntax-rules``, see https://docs.racket-lang.org/guide/pattern-macros.html#%28part._define-syntax_and_syntax-rules%29 .
```

Ericson2314 commented:
Updated to properly distinguish Scheme and Racket :)

nomeata commented:
@Ericson2314, are you by any chance attending Haskell Symposium? If so, would you like to present this proposal at the [GHC Proposals session](https://icfp22.sigplan.org/home/haskellsymp-2022#GHC-Proposals-Session)?

Ericson2314 commented:
@nomeata I unfortunately am not going to ICFP --- and feeling some regret right now for sure!

Ericson2314 commented:
Hmm I guess I forgot to submit this to the committee?

nomeata commented:
Looks like it :-). Is it ready fonr submission?

Ericson2314 commented:
@nomeata I think so!

adamgundry commented on proposals/0529-th-quotes-as-patterns.rst:
```diff
@@ -0,0 +1,356 @@
+  --------------------------------
+  [| ... $(p) ... |] is in <apat>
+
+The third and final difference is that names in quotes must all be uses, never bindings.
```
Presumably this means that matching on most top-level declarations is prohibited, as e.g. `[d| data T = MkT |]` binds names? Are there some examples where it is useful to match on a top-level declaration that does not bind any names?

I think you need to spell out the semantics of pattern-matching on name uses. I gather from the examples that you intend `[| name |]` to expand to a match on `VarE` plus an equality test with the in-scope `name`. But that should be written down explicitly as part of the Proposed Change Specification.

Ericson2314 replied:
I found 
```haskell
_ =...
```
as an example of a top level declaration that doesn't bind names
```haskell
instance $(t)
```
also

I will spell out.

adamgundry commented on proposals/0529-th-quotes-as-patterns.rst:
```diff
@@ -0,0 +1,356 @@
+Implementation Plan
+-------------------
+
+I lack the time capacity to implement these changes all by myself, and would submit this to the Haskell Foundation as part of whatever fund as part of whatever https://discourse.haskell.org/t/pre-hftt-ongoing-focus-on-migration-tools/4626 becomes.
```
```suggestion
I lack the time capacity to implement these changes all by myself, and would submit this to the Haskell Foundation as part of whatever fund https://discourse.haskell.org/t/pre-hftt-ongoing-focus-on-migration-tools/4626 becomes.
```

Ericson2314 replied:
```suggestion
I lack the time capacity to implement these changes all by myself, and would submit this to the Haskell Foundation to fund as part of whatever https://discourse.haskell.org/t/pre-hftt-ongoing-focus-on-migration-tools/4626 becomes.
```

adamgundry commented on proposals/0529-th-quotes-as-patterns.rst:
```diff
@@ -0,0 +1,356 @@
+
+- https://arxiv.org/pdf/2001.10490.pdf
+
+Future Work
```
Would it make sense to have something like this for Typed Template Haskell? How far off is such a thing?

Ericson2314 replied:
Don't think it is too bad, but not sure. Added as future work.

adamgundry commented on proposals/0529-th-quotes-as-patterns.rst:
```diff
@@ -0,0 +1,356 @@
+
+   It is disallowed because the first ``x`` in the quote is a binding not a use.
+
+Effect and Interactions
```
I'm assuming that the `TemplateHaskell` will not imply the new extension, though it does imply `TemplateHaskellQuotes`?

Ericson2314 replied:
Yes. Added.

adamgundry commented on proposals/0529-th-quotes-as-patterns.rst:
```diff
@@ -0,0 +1,356 @@
+                 implements the described feature.
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/529>`_.
+.. contents::
```
```suggestion
.. sectnum::
.. contents::
```

adamgundry commented on proposals/0529-th-quotes-as-patterns.rst:
```diff
@@ -0,0 +1,356 @@
+
+The second differences is that splices within these quotes contain patterns instead of expressions::
+
+  p is in <pat>
+  --------------------------------
+  [| ... $(p) ... |] is in <apat>
```
This looks at first glance like a typing judgement, but in fact is describing the grammar (only)?

adamgundry reviewed (changes requested):
Shepherd here. Thanks for taking the time to write up this proposal @Ericson2314! The motivation seems solid to me, but I think the specification could do with some additional detail. I've added detailed commentary inline.

Ericson2314 commented on proposals/0529-th-quotes-as-patterns.rst:
```diff
@@ -0,0 +1,356 @@
+
+In this way, we also avoid the use of the AST.
+
+Note we do have quotes *of patterns* today (``[p| ... |]``), but that is orthogonal.
+This is quotes *as patterns*, the type of syntax being quoted doesn't matter and could be anything.
+The point is the quotes are in negative position.
+
+With this change put together, the hope is that a significant portion of TH out in the wild is going to be more stable across GHC versions.
+This is all accomplished without trying to minimize TH AST changes, which is quite a hopeless task and also a perverse incentive for the rest of language development.
```
I wish, but I am afraid the equivalent of
```haskell
f [qq| ... |] = ...
```
is just
```haskell
f $(ff "...") = ...
```
i.e. the "regular" splice of pattern syntax we have today.

Ericson2314 commented on proposals/0529-th-quotes-as-patterns.rst:
```diff
@@ -0,0 +1,356 @@
+
+With the new extension ``TemplateHaskellQuotesAsPatterns``, slightly modified quotes are usable in pattern position.
+
+The first difference is that quotes as pattern match raw syntax, not (monadic) actions producing syntax.
+The tying rules are as follows:
```
I was worried "quotes in patterns" would sound like `f $(... [| ... |] ...) = ...` but maybe I am overthinking it, and people will realize that "quotes in patterns" is not the same as "quotes in splices in patterns"?

adamgundry replied:
Right, I see that it is hard to have unambiguous terminology here. That's why I think the proposal should explicitly define its terms (https://github.com/ghc-proposals/ghc-proposals/pull/529#discussion_r1032582486). Having some definitions and using them consistently is more important than the precise choice of words in those definitions, I think.

Ericson2314 commented on proposals/0529-th-quotes-as-patterns.rst:
```diff
@@ -0,0 +1,356 @@
+
+`Proposal 246`_ made it so that quotes are polymorphic, e.g. ``[| 1 + 1 |] :: Quote m => m Exp``.
+The ``Quote`` class has a ``newName`` method, and is just used when binding local variables.
+Relax the rules so that TH Quotes only impose a ``Quote`` constraint when ``newName`` is in fact needed.
```
OK added a bunch of text and cross-referencing around these things. Hope this is good now!

Ericson2314 commented on proposals/0529-th-quotes-as-patterns.rst:
```diff
@@ -0,0 +1,356 @@
+
+  [| a + $(x) |]
+
+could match any ``t Exp`` where ``t`` is an bind ``x :: t Expr`` with this desugaring::
```
Indeed. Sorry for all the typos! :grimacing:

Ericson2314 commented on proposals/0529-th-quotes-as-patterns.rst:
```diff
@@ -0,0 +1,356 @@
+It is also unclear whether the ``nameBase _ == "x"`` is useful in practice, or whether it is better to just "bake in" alpha equivalence and not care whether the local variable is an "x" or not provided the usage lines up with the binding.
+
+Finally, the non-linear patterns trick is not a true dual because it merely checks whether the variables "ended up" being the same *once the action is run*, rather than pattern matching on the action *itself*.
+De Bruijn indices encode actions in a way that makes equality of the easily decidable, for example, and thus would be a less hacky solution.
```
Done.

Ericson2314 commented on proposals/0529-th-quotes-as-patterns.rst:
```diff
@@ -0,0 +1,356 @@
+
+None at this time.
+
+Related Work
```
I am frankly not sure I am expert enough either.

GHC is now quite beyond the state of the art in hygiene theory (probably Nominal Sets) or practice (probably Racket).

I am trying hard to straight-jacket the system so that the underlying lack of binding structure doesn't shine through, but it isn't entirely clear to me to what extend my efforts succeed.

`[A]X` Nominal set name bindings can I believe be related to `Quote`, in that  it should be possible to log all the "new names" and then establish an equivalence class the term with permutations of them, but:

- Once we run the action we loose that issue
- We will loose the *internal* binding structure because this is not compositional, with subterms having their own post-hoc reconstructable actions. Monad-ing everything together looses that information.

Ericson2314 replied:
I just stumbled upon https://www.mail-archive.com/racket-users@googlegroups.com/msg37427.html from @lexi-lambda, so I would say and she and @david-christiansen would both be good people with a foot in Haskell and Racket land to help expand this section.

Ericson2314 replied:
https://srfi.schemers.org/srfi-72/srfi-72.html is not quite formal, but interesting. I think @lexi-lambda in that thread and me reading now but found it appealing for much the same reasons.

sgraf812 replied:
Correct me if I'm wrong, but since the proposal currently does not allow matching on binding constructs, we can't make hygiene better or worse. Judging from an hour worth reading, I also don't think we lock ourselves out with this proposal to more modern approaches such as `syntax-parse` because the stated semantics should be a subset of whatever advanced stuff we will happen to want in the future.

int-index commented:
As far as terminology is concerned, I'd like to suggest to make a distinction between "pattern quotes" and "quote patterns".

**Pattern quote**: a Template Haskell quote that constructs a pattern, i.e. `[p| ... |]`
**Quote pattern**: a pattern that consists of a Template Haskell quote (i.e. the feature proposed here)

Based on this terminology, `TemplateHaskellQuotesAsPatterns` could be renamed to `TemplateHaskellPatternQuotes`.

Ericson2314 commented:
I like that, @int-index. But moreover I would love to wash my hands of all name bike-shedding, purely following the consensus of others on this one :). So I hope you and @adamgundry can come to a decision on that!

Ericson2314 commented:
@adamgundry I believe I have addressed most of the issues, except for the naming issue, and expanding the Related Work section.

Perhaps some sort of interactive discussion in what sort of topics would be useful to discuss, so I know what I should delve into.

Maybe a good starting point is also that the Scheme family macros work fundamentally differently, in that that the binding structure is unknown because the syntax is sexpression of unknown meaning until the final splice. Whereas we quote actual Haskell expressions and thus the quote-internal binding structure *is* known from the get-go, before manipulation. Perhaps therefore the most important one to investigate is Rust, since it is very Racket inspired but does allow quotes of expression etc. (rather than mere "token tree", it's s-expression equivalent) like us.

I am not at all familiar with what Idris and Lean do.

simonpj commented:
I get the general motivation but lots of things are unclear here.

1. Is this an extension to *untyped* Template Haskell, or to *typed* Template Haskell?
   I assume the former, but it would be much better to say so.  The important point is that
   * Untyped TH splices can appear in patterns (and declarations and types) whereas typed TH splices are always expressions..
   * Untyped TH splices run during *renaming*, whereas typed TH spliced run during *type checking*.

   If it's untyped TH, then it doesn't make so much sense to talk about the typing rules for it.
   All we need to know is that any nested splices `$p` must be for patterns `p`.

2. I'd love the relationship to quasi-quotes to be far more explicit.  [Quasi-quotations](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/template_haskell.html?highlight=quasiquote#extension-QuasiQuotes) (`-XQuasiQuotes`) are
   essentially untyped Template Haskell.  The quasi-quote `[foo| string |]` in an
   pattern context is precisely equvalent to `$(quotePat foo "string")`, where `quotePat` picks a field from the record `foo :: QuasiQuoter`.
   ```
   data QuasiQuoter = QuasiQuoter { quoteExp  :: String -> Q Exp,
                                    quotePat  :: String -> Q Pat,
                                    quoteType :: String -> Q Type,
                                    quoteDec  :: String -> Q [Dec] }
   ```
   This splice runs during *renaming*.  For example:
   ```
   f [wombat| hello %x there |] = x+2
   ```
   here the quasiquote `[wombat| hello %x there |]` expands to a pattern that (presumably) binds `x`.  But the syntax for this binding is entirely under the control of the `wombat` quasiquoter, which can do whatever it likes.

3. Currently you can't use `e`, `p`, `d`, or `t` as the name of a quasiquoter.
   * In expression position `[e| blah|]` is an expression quote of type `Q Exp`, and `[t| blah|]` is a type quote of type `Q Typ`.
   * These forms are simply not allowed in pattern position, in types etc.

   So the proposal plans to make them meaningful in patterns.  Well and good, but say so much more explicitly.  It's really as if we are providing a rather special quasi-quoter `e` for which `quotePat e` does the thigns described here.  Quite nice: fills out out a corner in the design.   

4. There is an awkwardness about binding. Consider
   ```
   f n = blah

   h x = [e| Just $(f 2) |]

   g [e| ($x,  $(f 2)) |] = ...
   ```
   In a normal expression splice, like in the RHS of `h`, the `f` is a *bound* occurrence of `f`, and the splice `$(f 2)` is run to compute the syntax tree to plug in instead of `$(f 2)`.

   But what about in patterns?  I think the `$x` is a *binding* occurrence; and I think
   that the splice `$(f 2)` is perhaps illegal, *because it is understood as a pattern*.
   (I think that is what the crypic statement "The second difference is that splices within these quotes contain patterns instead of expressions" means.)

   That means that splices in pattern-quotes can do no computation (unlike splices in expression-quotes).  Maybe that's ok.

5. "The third and final difference is that names in quotes must all be uses, never bindings." I have no idea what this means.

adamgundry commented:
@simonpj 

> 5. "The third and final difference is that names in quotes must all be uses, never bindings." I have no idea what this means.

The examples point out that binding a name inside a quote-as-pattern is disallowed, so this is rejected because of the binding occurrence of `x`:
```hs
f [| \x -> x |] = ...
```

There is a bit more discussion of this in the Alternatives, section 6.2. The point is that picking the right semantics for such bindings seems nontrivial, so the proposal defers committing to a semantics.

adamgundry commented on proposals/0529-th-quotes-as-patterns.rst:
```diff
@@ -0,0 +1,356 @@
+Quotes as patterns
+~~~~~~~~~~~~~~~~~~
+
+With the new extension ``TemplateHaskellQuotesAsPatterns``, slightly modified quotes are usable in pattern position.
```
I see that @int-index also had a suggestion in https://github.com/ghc-proposals/ghc-proposals/pull/529#issuecomment-1340406138, although my inclination is still to have explicit prepositions rather than relying on word order to distinguish "pattern quote" from "quote pattern". I'm happy to leave the bikeshed colour to be resolved later once the substance of the proposal is agreed, though.

I still think it would be helpful to have some explicit definitions, something like (with the current terminology):

 * A quote is said to be *of* the syntactic category being quoted, for example `[e| ... |]` is a *quote of an expression* and `[p| ... |]` is a *quote of a pattern*. Template Haskell already supports quotes of expressions, patterns, types and declarations, plus quotes of arbitrary syntax via quasiquotes.
 * A quote is said to be used *as* the syntactic category in which it occurs, for example `x = [...| ... |]` includes a  *quote as an expression* whereas `f [...| ... |] = ...` includes a *quote as a pattern*. Template Haskell currently supports quotes as expressions and quasiquotes as all four categories; this proposal adds quotes as patterns.
 * Thus in `f [t| Bool |] = [d| data T |]` we have a *quote of a type as a pattern* and a *quote of a declaration as an expression*.

nomeata commented:
Should we mark this as “needs revision”?

Ericson2314 commented:
Yes please. And I had written it for HF consideration but it is now pretty far down in the future work on the draft proposal I am working on so it feels less urgent (perhaps I shouldn't have submitted it to the committee).

I am glad I wrote it, I am thankful @adamgundry reviewed it, but I feel less urgency pushing it to completion until the lower hanging fruit we can do re `template-haskell` compatibility (e.g. record fields as @bgamari pointed out) is plucked. I still think it is a good idea, but as the saying goes "walk before you run".

adamgundry commented:
Thanks for this @Ericson2314, I'm happy to label this as "needs revision" for now and wait for you or another willing volunteer to decide to pick it up and get it over the line. As I indicated I think this is basically a solid idea, albeit a nontrivial one to implement and with a few interesting corners to discuss. :smile:

sgraf812 commented:
As we discussed on Sunday ([Ben's notes](https://edit.smart-cactus.org/u__IGA1bTd2DpulYmlnxaw)) and as the proposal states, writing such "quote patterns" is a great way for users that use TH for the macros/grammar use case (as opposed to the staging/code gen use case) to define their macros in a way that is far less susceptible to churn in the TH AST, plus it is far more ergonomically to use. Perhaps the discussion on Sunday would re-ignite @Ericson2314's interest on the matter?

As far as bike-shedding is concerned, I'd like to throw "Quotes in matches" or "matching quotes" into the ring. That expresses that these things must occur in a meta level pattern match, in "negative" position. There is no risk of confusing that with "Quote patterns" or "pattern quotes" where which is which I already forgot since I started this post. Unfortunately, a match is already a name describing the syntactic construct that encompasses a pattern and its GRHSs so that name isn't optimal either.

Perhaps we should take strong inspiration in https://docs.racket-lang.org/guide/pattern-macros.html. Example

```hs
foo :: Int -> Int
f [| foo (2*$(x)) |] = [| 2 * foo $(x) |]
```

(I hope I got that right.) Here, `[| foo (2*$(x)) |]` is called the *pattern* and `[| 2 * foo $(x) |]` is called the *template*. Both the pattern and the template use *quotes* which have different semantics based on where they appear. A template quote that produces a pattern (with the `[p|..]` quoter) is separate from a pattern quote, which is a meta-level pattern taking the syntactic shape of a quote. As the proposal states, pattern quotes match on pure AST expressions whereas template quotes have monadic type.

Template Haskell used to have template quotes only, now it also got pattern quotes. (Makes you wonder if the name still flies.)

I don't think it makes sense to call `f` a *pattern-based macro*, because that term would apply to `foo` if `f` were automatically applied everywhere in the program; we currently lack such automated extension of the meta-level grammar.

Kha commented on proposals/0529-th-quotes-as-patterns.rst:
```diff
@@ -0,0 +1,513 @@
+Support local variables
+~~~~~~~~~~~~~~~~~~~~~~~
+
+Quotes that bind local variables do in fact have an interpretation as non-linear patterns::
```
In macro systems like Racket's and Lean's, you cannot distinguish between bindings and uses in quotations because scoping is discovered only during expansion. So "quotes that bind local variables" is not even something we can detect in those systems.

sgraf812 replied:
I presume "discovered only during expansion" is an artifact of the fact that  Racket-style macros will not get a name-resolved TH AST as input (where we know the binding structure but limit ourselves to macro syntax expressible in Haskell) but rather just a concrete syntax tree. Perhaps let me rephrase in terms of the example in https://school.racket-lang.org/2019/plan/tue-mor-lecture.html:

```
(define-syntax (robust-for/list5 stx)
  (syntax-parse stx
    [(_ ([elem-name:id seq]) computation)
     #'(map (λ (elem-name) computation) seq)]))
```

This macro will expand `(robust-for/list5 ([x (list 1 2)]) (add1 x))` to `(map (λ (x) (add1 x)) (list 1 2))`, matching `[elem-name :-> x, seq :-> (list 1 2)]`. Note that `elem-name` is ultimately spliced into a binding occurrence and that `computation` may refer to the instantiation of `elem-name`, thus its *syntax class* is specified to be `id`, which is what `elem-name:id` does. (This is somewhat like pushing an expected type for the AST node to match inside the pattern, I guess, where "type" in Racket really means a user-definable non-terminal of the extensible syntax. See 5.6 in the link above.) But the *pattern* `(_ ([elem-name:id seq]) computation)` does not know about binding structure at all; rather the macro equips the pattern with a binding structure post-hoc when it generates the template `(map (λ (elem-name) computation) seq)]))` for the pattern, thus intentionally capturing occurrences of `elem-name` in `computation`.

Kha replied:
That's exactly right. It's the same reason I mentioned in the ZuriHac discussion why our macros are untyped (on the level of the quoted language) but core-level transformations can optionally be typed (which is done by https://github.com/gebner/quote4 btw).
